### PR TITLE
Add generic document management in editoast

### LINKED ISSUE #3344: Create a generic endpoint to get an image (Studies and Rolling Stock)
## Description

Add a generic system to store and retrieve documents and images.

## Acceptance criteria

- Have an endpoint to add a document (given an optional key)
  - If not given, use the hash of the document
  - Return the key
  - If a document exists return a 400 error
- Add an endpoint to retrieve a document
- Add an endpoint to delete a document
- Have functions to add, delete, and get the URL of a document.

### PR DESCRIPTION
**Pull request by flomonster**
The idea of this PR is to add support for uploading documents such as images, pdf and so on. 

We have two immediate uses, one for rolling stock liveries and another one for project images.
In API each usage has its own implementation and way to store the data. With the `documents` module I've added we can now add, retrieve, and delete documents easily with a key system.

The developer can either choose to specify a key manually for their documents or use hashable objects.

### CODE REVIEW

File: editoast/src/documents.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+/// **WARNING**: The key of a document must be unique.
+///
+/// ## Example
+///
+/// Let's say we have an image of a study to save in the database.
+/// A way to construct your unique key can be to use a tuple of the study id and a static `str`.
+///
+/// ```
+/// let study_pk = 42;
+/// let key = DocumentKey::new_hash(&format("study_image_{study_pk}"));
```
Comment (multun) [inline editoast/src/documents.rs]: can't we automatically hash the content and use that as a key? (content-addressing)
if that is not the goal, then what is the key hash used for?

File: editoast/src/documents.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+impl DocumentKey {
+    /// Create a document key from a string
+    pub fn new<T: AsRef<str>>(key: T) -> Self {
+        Self(key.as_ref().into())
+    }
+
+    /// Create a document key from any data hashable object
+    #[allow(dead_code)] // TODO: Remove once it's used by studies or rolling stocks
+    pub fn new_hash<T: AsRef<[u8]>>(hashable_objects: &T) -> Self {
+        let mut hasher = Sha1::new();
```
Comment (multun) [inline editoast/src/documents.rs]: is it intended to be a cryptographic hash?

File: editoast/src/documents.rs (hunk 1 above)
Comment (younesschrifi) [inline editoast/src/documents.rs]:
```suggestion
/// let project_pk = 42;
/// let key = DocumentKey::new_hash(&format("project_image_{project_pk}"));
```

**Review by younesschrifi — CHANGES_REQUESTED** [review]

File: editoast/src/documents.rs (hunk 1 above)
Comment (flomonster) [reply editoast/src/documents.rs]: In some cases, we don't want to do that. For example for project images. We will probably have a foreign key to the document so when a project is deleted its image is deleted in cascade. This doesn't work if the key used is the hash of the image (when two projects have the same image).

Using the hash of the document as the key will probably be used for liveries.

File: editoast/src/documents.rs (hunk 2 above)
Comment (flomonster) [reply editoast/src/documents.rs]: I don't especially need a cryptographic hash. I just need a hashing algorithm that has strong collision resistance. Sha1 has this property and is fast to compute. Do you have another algorithm in mind?

**Review by younesschrifi — DISMISSED** [review]

**Review by multun — APPROVED** [review]

**Review by younesschrifi — APPROVED** [review]

**Pull request merged**